Directional contact interaction

yutakasi634 opened the pull request:
- implement directional contact interaction with tests.
- implement uniform potential with reading test.

yutakasi634 committed:
chore:make input file example for directional contact

yutakasi634 committed:
feat: implementing DirectionalContactInteraction

yutakasi634 committed:
feat: add read_directional_contact_potential function

yutakasi634 committed:
feat: add read_directional_contact_interaction

yutakasi634 committed:
feat: implement calc force in DirectionalContactInteraction

yutakasi634 committed:
feat: implement calc energy in DirectionalContactInteraction

yutakasi634 committed:
fix: some bug in DirectionalContactInteraction

yutakasi634 committed:
test: implementing test for directional contact interaction

yutakasi634 committed:
update toml11 version

yutakasi634 committed:
test: add test for read directional contact interaction

yutakasi634 committed:
doc:write book about directional contact interaction

yutakasi634 committed:
feat: add Gaussian option to angle of directional contact interaction

yutakasi634 committed:
feat: change return value of DirectionalContactInteraction::name()

yutakasi634 committed:
fix: permutation of angle potential order in read local interaction

yutakasi634 committed:
fix: return locate in unexpected position

yutakasi634 committed:
fix: bug in Directional contact calc_force

ToruNiina commented on book/ja/refs/DirectionalContactInteraction.md:
```diff
@@ -0,0 +1,54 @@
+margin             = 0.5 # relative length to longest cutoff
+parameters         = [
+    {indices = [0, 1, 2, 3], ... }, angle1_params = { v0 = ... }, angle2_params = { v0 = ... }, contact_params = { v0 = ... }}, # 各ポテンシャルそれぞれについて、要求されるパラメータを設定します。
+    必要に応じて続きます...
```
```suggestion
    # 必要に応じて続きます...
```

ToruNiina commented on book/ja/refs/DirectionalContactInteraction.md:
```diff
@@ -0,0 +1,54 @@
+
+`DirectionalContact`相互作用を用いる場合、`[[forcefields.local]]`テーブルには、以下の値を設定します。
+
+- `interactin`: 文字列型
```
```suggestion
- `interaction`: 文字列型
```

ToruNiina commented on book/ja/refs/DirectionalContactInteraction.md:
```diff
@@ -0,0 +1,54 @@
+topology           = "none"
+margin             = 0.5 # relative length to longest cutoff
+parameters         = [
+    {indices = [0, 1, 2, 3], ... }, angle1_params = { v0 = ... }, angle2_params = { v0 = ... }, contact_params = { v0 = ... }}, # 各ポテンシャルそれぞれについて、要求されるパラメータを設定します。
```
```suggestion
    # 各ポテンシャルそれぞれについて、要求されるパラメータを設定します。
    {indices = [0, 1, 2, 3], angle1_params = { v0 = ... }, angle2_params = { v0 = ... }, contact_params = { v0 = ... }},
```

ToruNiina commented on book/ja/refs/DirectionalContactInteraction.md:
```diff
@@ -0,0 +1,54 @@
+- `potentials`
+  - `angle1`,`angle2`: 文字列型
+    - ["Cosine"](Cosine.md): コサイン関数を使った汎用的なポテンシャルです。
+  - `contact`:
```
```suggestion
  - `contact`: 文字列型
```

ToruNiina commented on book/ja/refs/DirectionalContactInteraction.md:
```diff
@@ -0,0 +1,54 @@
+`DirectionalContact`相互作用は、結合が形成される粒子間の距離とその結合方向に対してかかる相互作用です。
+ポテンシャルは角度に関するポテンシャルと距離に関するポテンシャルの積として表現されます。
+
+粒子`i`番目,`j`番目,`k`番目,`l`番目の粒子について、`i`,`j`,`k`がなす角に`angle1`で指定したポテンシャルが、`j`,`k`,`l`がなす角に、`angle2`で指定したポテンシャルが、`j`,`k`間の距離に`contact`で指定したポテンシャルがかかります。
```
```suggestion
粒子`i`番目,`j`番目,`k`番目,`l`番目の粒子について、`i`,`j`,`k`がなす角に`angle1`で指定したポテンシャルが、`j`,`k`,`l`がなす角に`angle2`で指定したポテンシャルが、`j`,`k`間の距離に`contact`で指定したポテンシャルがかかります。
```

ToruNiina commented on input/directional_contact.toml:
```diff
@@ -0,0 +1,73 @@
+
+[[systems]]
+attributes.temperature = 300.0
+boundary_shape.lower = [ 0.0,    0.0,   0.0]
```
```suggestion
boundary_shape.lower = [  0.0,   0.0,   0.0]
```

ToruNiina commented on mjolnir/input/read_local_interaction.hpp:
```diff
@@ -122,6 +123,136 @@ read_contact_interaction(const std::string& kind, const toml::value& local)
+// This is internal function for read_directional_contact_interaction function.
+template<typename traitsT, typename angle1_potentialT, typename angle2_potentialT>
+typename std::unique_ptr<LocalInteractionBase<traitsT>>
+read_contact_in_directional_contact_interaction(const std::string& kind, const toml::value& local)
```
Please, for consistency to other functions, use 4-spaces indentation.
All other functions that are added to this file also.

ToruNiina commented on mjolnir/input/read_local_interaction.hpp:
```diff
@@ -122,6 +123,136 @@ read_contact_interaction(const std::string& kind, const toml::value& local)
+  MJOLNIR_LOG_FUNCTION();
+  std::vector<std::string> angle_potential_keys = {"angle2", "angle1"};
+
+  return read_angle_in_directional_contact_interaction<traitsT>(kind, local, angle_potential_keys);
```
If you use non-const value, it is good to move it here.
```cpp
return read_angle_in_directional_contact_interaction<traitsT>(kind, local, std::move(angle_potential_keys));
```

If you don't like move-semantics for some reason, mark is `const` because it will not be modified.

```cpp
const std::vector<std::string> angle_potential_keys = {"angle2", "angle1"};
```

IMHO, the simplest implementation here is the following.
```cpp
  return read_angle_in_directional_contact_interaction<traitsT>(kind, local, {"angle2", "angle1"});
```

ToruNiina commented on mjolnir/input/read_local_interaction.hpp:
```diff
@@ -122,6 +123,136 @@ read_contact_interaction(const std::string& kind, const toml::value& local)
+  using real_type = typename traitsT::real_type;
+
+  std::string angle_potential_key = angle_potential_keys.back();
+  std::string angle_potential = toml::find<std::string>(local, "potentials", angle_potential_key);
```
Here, the types are enough clear. You can use `auto`.
Also, you must use `const` because these values will never be modified. You need to avoid accidental modification on those variables.
```suggestion
  const auto angle_potential = toml::find<std::string>(local, "potentials", angle_potential_key);
```

ToruNiina commented on mjolnir/input/read_local_potential.hpp:
```diff
@@ -175,6 +176,19 @@ read_cosine_potential(const toml::value& param, const toml::value& env)
+
+  auto k = find_parameter<real_type>(param, env, "k");
+
+  MJOLNIR_LOG_INFO("UniformPotential");
```
```suggestion
  MJOLNIR_LOG_INFO("UniformPotential = {k = ", k, '}');
```

ToruNiina commented on mjolnir/input/read_local_potential.hpp:
```diff
@@ -175,6 +176,19 @@ read_cosine_potential(const toml::value& param, const toml::value& env)
 
+template<typename realT>
+UniformPotential<realT>
+read_uniform_potential(const toml::value& param, const toml::value& env)
```
Use 4-space indentation for the consistency sake.

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+#ifndef MJOLNIR_INTERACTION_DIRECTIONAL_CONTACT_INTERACTION_HPP
```
```suggestion
#ifndef MJOLNIR_INTERACTION_LOCAL_DIRECTIONAL_CONTACT_INTERACTION_HPP
```

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+#ifndef MJOLNIR_INTERACTION_DIRECTIONAL_CONTACT_INTERACTION_HPP
+#define MJOLNIR_INTERACTION_DIRECTIONAL_CONTACT_INTERACTION_HPP
```
```suggestion
#define MJOLNIR_INTERACTION_LOCAL_DIRECTIONAL_CONTACT_INTERACTION_HPP
```

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+         typename contact_potentialT>
+class DirectionalContactInteraction final : public LocalInteractionBase<traitsT>
+{
+ public:
```
4-space indentation.

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+      {
+        this->active_contacts_.push_back(i);
+      }
+      this->current_margin_ = this->cutoff_ * this->margin_;
```
`this->cutoff_ * this->margin_` does not change while this for loop. I don't think you need to re-write `this->current_margin_` every time. Move this out from the loop.

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+    for(std::size_t i=0;i < this->potentials_.size(); ++i)
+    {
+      const indices_potentials_tuple& pot = this->potentials_[i];
+      const coordinate_type pos0 = sys.position(std::get<0>(pot)[1]);
```
```suggestion
      const coordinate_type& pos0 = sys.position(std::get<0>(pot)[1]);
```

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+    {
+      const indices_potentials_tuple& pot = this->potentials_[i];
+      const coordinate_type pos0 = sys.position(std::get<0>(pot)[1]);
+      const coordinate_type pos1 = sys.position(std::get<0>(pot)[2]);
```
```suggestion
      const coordinate_type& pos1 = sys.position(std::get<0>(pot)[2]);
```

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+    // absolute length of margin (this->margin_ is a relative length).
+    const real_type abs_margin = this->cutoff_ * this->margin_;
+
+    for(std::size_t i=0;i < this->potentials_.size(); ++i)
```
```suggestion
    for(std::size_t i=0; i<this->potentials_.size(); ++i)
```

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+  MJOLNIR_GET_DEFAULT_LOGGER_DEBUG();
+  MJOLNIR_LOG_FUNCTION_DEBUG();
+
+  for(const std::size_t active_contact : this -> active_contacts_)
```
```suggestion
  for(const std::size_t active_contact : this->active_contacts_)
```

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+  {
+    const auto& idxp = this->potentials_[active_contact];
+
+    const auto angle1_pot  = std::get<1>(idxp);
```
```suggestion
    const auto& angle1_pot  = std::get<1>(idxp);
```

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+    const auto& idxp = this->potentials_[active_contact];
+
+    const auto angle1_pot  = std::get<1>(idxp);
+    const auto angle2_pot  = std::get<2>(idxp);
```
```suggestion
    const auto& angle2_pot  = std::get<2>(idxp);
```

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+
+    const auto angle1_pot  = std::get<1>(idxp);
+    const auto angle2_pot  = std::get<2>(idxp);
+    const auto contact_pot = std::get<3>(idxp);
```
```suggestion
    const auto& contact_pot = std::get<3>(idxp);
```

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+
+    const coordinate_type  Pij = sys.adjust_direction(rPj - rPi); // Pi -> Pj
+    const real_type       lPij = math::length(Pij);
+    if(lPij > contact_pot.cutoff())
```
Why don't you compare `length_sq` and `cutoff() * cutoff()` here?

yutakasi634 replied:
After here, I use lPij to calculate inv_len_Pij, so in any case, I  have to calculate lPij, and this is the reason why calculating lPij here.

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+    const real_type inv_len_PjCj        = math::rlength(PjCj);
+    const coordinate_type PjCj_reg      = PjCj * inv_len_PjCj;
+
+    const coordinate_type Pji_reg       = - Pij_reg;
```
Be consistent. You (and other codes currently in Mjolnir) don't have any space between `operator-` and its operand. But here you have one.
```suggestion
    const coordinate_type Pji_reg       = -Pij_reg;
```

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+    // dU_con(|Pij|) / dr
+    const real_type contact_coef      = contact_pot.derivative(lPij);
+    const coordinate_type dU_con_drPj = contact_coef * Pij_reg;
+    const coordinate_type dU_con_drPi = - dU_con_drPj;
```
```suggestion
    const coordinate_type dU_con_drPi = -dU_con_drPj;
```

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+  {
+    const auto& idxp = this->potentials_[active_contact];
+
+    const auto angle1_pot  = std::get<1>(idxp);
```
```suggestion
    const auto& angle1_pot  = std::get<1>(idxp);
```

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+    const auto& idxp = this->potentials_[active_contact];
+
+    const auto angle1_pot  = std::get<1>(idxp);
+    const auto angle2_pot  = std::get<2>(idxp);
```
```suggestion
    const auto& angle2_pot  = std::get<2>(idxp);
```

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+
+    const auto angle1_pot  = std::get<1>(idxp);
+    const auto angle2_pot  = std::get<2>(idxp);
+    const auto contact_pot = std::get<3>(idxp);
```
```suggestion
    const auto& contact_pot = std::get<3>(idxp);
```

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+
+    const coordinate_type Pij = sys.adjust_direction(rPj - rPi); // Pi -> Pj
+    const real_type lPij = math::length(Pij);
+    if(lPij > contact_pot.cutoff())
```
Why don't you compare `length_sq` and `cutoff * cutoff`?

yutakasi634 replied:
Same as above.

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+    const real_type inv_len_PjCj   = math::rlength(PjCj);
+    const coordinate_type PjCj_reg = PjCj * inv_len_PjCj;
+
+    const coordinate_type Pji_reg = - Pij_reg;
```
```suggestion
    const coordinate_type Pji_reg = -Pij_reg;
```

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+         typename angle2_potentialT, typename contact_potentialT>
+void DirectionalContactInteraction<traitsT, angle1_potentialT,
+                                   angle2_potentialT, contact_potentialT>
+::write_topology(topology_type& topol) const
```
```cpp
/*...*/ <
// ...
>::write_topology
```

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+  if(this->kind_.empty() || this->kind_ == "none") {return;}
+
+  for(const auto& idxp : this->potentials_)
+    {
```
Indentation.

ToruNiina commented on mjolnir/potential/local/UniformPotential.hpp:
```diff
@@ -0,0 +1,49 @@
+#ifndef MJOLNIR_POTENTIAL_UNIFORM_POTENTIAL_HPP
```
```suggestion
#ifndef MJOLNIR_POTENTIAL_LOCAL_UNIFORM_POTENTIAL_HPP
```

ToruNiina commented on mjolnir/potential/local/UniformPotential.hpp:
```diff
@@ -0,0 +1,49 @@
+#ifndef MJOLNIR_POTENTIAL_UNIFORM_POTENTIAL_HPP
+#define MJOLNIR_POTENTIAL_UNIFORM_POTENTIAL_HPP
```
```suggestion
#define MJOLNIR_POTENTIAL_LOCAL_UNIFORM_POTENTIAL_HPP
```

ToruNiina commented on mjolnir/potential/local/UniformPotential.hpp:
```diff
@@ -0,0 +1,49 @@
+class UniformPotential
+{
+ public:
+  using real_type = realT;
```
indentation.

ToruNiina commented on mjolnir/potential/local/UniformPotential.hpp:
```diff
@@ -0,0 +1,49 @@
+
+ public:
+  UniformPotential(const real_type k) noexcept
+      : k_(k), cutoff_(std::numeric_limits<real_type>::max())
```
If `cutoff_` is a constant and will never be changed, is there any reason to have it as a member variable? It consumes memory resource, but does nothing.
Also, here it should be `infinity()` because the cutoff distance is not a finite number.

ToruNiina commented on mjolnir/potential/local/UniformPotential.hpp:
```diff
@@ -0,0 +1,49 @@
+  static const char* name() noexcept {return "Uniform";}
+
+  real_type k()      const noexcept {return k_;}
+  real_type cutoff() const noexcept {return cutoff_;}
```
You can return `std::numeric_limits<real_type>::infinity()` instead of member variable. Remember, when you add a member variable, you will consume more memory resource.

ToruNiina commented on mjolnir/potential/local/UniformPotential.hpp:
```diff
@@ -0,0 +1,49 @@
+ private:
+
+  real_type k_;
+  real_type cutoff_;
```
For the reasons described above, remove this.

ToruNiina commented on test/core/test_directional_contact_interaction.cpp:
```diff
@@ -0,0 +1,230 @@
+
+BOOST_AUTO_TEST_CASE(DirectionalContactInteraction_numerical_diff)
+{
+  mjolnir::LoggerManager::set_default_logger("test_directional_contact_interaction.log");
```
indentation.

ToruNiina commented on test/core/test_directional_contact_interaction.cpp:
```diff
@@ -0,0 +1,230 @@
+  using directional_contact_type = mjolnir::DirectionalContactInteraction<
+    traits_type, angle_potential_type, angle_potential_type, contact_potential_type>;
+
+  const real_type pi = mjolnir::math::constants<real_type>::pi();
```
```suggestion
  constexpr real_type pi = mjolnir::math::constants<real_type>::pi();
```

ToruNiina commented on test/core/test_directional_contact_interaction.cpp:
```diff
@@ -0,0 +1,230 @@
+  const real_type angle1_native(pi);
+  const real_type angle2_native(pi);
+  const real_type contact_native(2.0);
+  angle_potential_type angle1_potential    = angle_potential_type{k_angle, 1, angle1_native};
```
Is there any reason to keep it non-const?
It is a test, so you need to ensure that accidental changes will not occur after here.
If you don't have any reason, add const.
Also, you don't need to write `angle_potential_type` twice.
```suggestion
const auto angle1_potential = angle_potential_type{k_angle, 1, angle1_native};
```

ToruNiina commented on test/core/test_directional_contact_interaction.cpp:
```diff
@@ -0,0 +1,230 @@
+  angle_potential_type angle2_potential    = angle_potential_type{k_angle, 1, angle2_native};
+  contact_potential_type contact_potential = contact_potential_type{k_contact, contact_native};
+  directional_contact_type interaction("none",
+      {{std::make_tuple(std::array<std::size_t, 4>{0,1,2,3},
```
```suggestion
      {{std::make_tuple(std::array<std::size_t, 4>{{0,1,2,3}},
```

ToruNiina replied:
Because `std::array` does not have any constructor, it is actually an aggregate type.
You need to use aggregate initialization here. You need to pass one array `{0, 1, 2, 3}` to a struct that has one array inside as an argument of initializer. So the first (outer) `{}` means initializer, and the next (inner) `{0, 1, 2, 3}` means temporary array to be passed.

ToruNiina commented on test/core/test_directional_contact_interaction.cpp:
```diff
@@ -0,0 +1,230 @@
+  sys.mass(1) = 1.0;
+  sys.mass(2) = 1.0;
+  sys.mass(3) = 1.0;
+  sys.at(0).position = {0.0, 0.0, 0.0};
```
```suggestion
  sys.position(0) = make_coordinate<coordinate_type>(0, 0, 0);
```

ToruNiina commented on test/core/test_directional_contact_interaction.cpp:
```diff
@@ -0,0 +1,230 @@
+        {
+          for(std::size_t k = 1; k <= contact_step_num; ++k)
+            {
+              real_type theta1 = i * dtheta;
```
const?

ToruNiina commented on test/core/test_directional_contact_interaction.cpp:
```diff
@@ -0,0 +1,230 @@
+          for(std::size_t k = 1; k <= contact_step_num; ++k)
+            {
+              real_type theta1 = i * dtheta;
+              real_type theta2 = j * dtheta;
```
const?

ToruNiina commented on test/core/test_directional_contact_interaction.cpp:
```diff
@@ -0,0 +1,230 @@
+                continue;
+              }
+
+              real_type r = k * dr;
```
const?

ToruNiina commented on test/core/test_read_directional_contact_interaction.cpp:
```diff
@@ -0,0 +1,101 @@
+    BOOST_TEST(angle2_pot.v0()  == real_type(0.0), boost::test_tools::tolerance(tol));
+    BOOST_TEST(angle2_pot.k()   == real_type(-20.0), boost::test_tools::tolerance(tol));
+    BOOST_TEST(angle2_pot.n()   == std::int32_t(2
+));
```
What happened?

ToruNiina commented on mjolnir/input/read_local_interaction.hpp:
```diff
@@ -122,6 +123,136 @@ read_contact_interaction(const std::string& kind, const toml::value& local)
 
+// This is internal function for read_directional_contact_interaction function.
+template<typename traitsT, typename angle1_potentialT, typename angle2_potentialT>
+typename std::unique_ptr<LocalInteractionBase<traitsT>>
```
`typename` is used to declare that a dependent qualified name (such as, `T::type`) is a type. Here, you don't need to add it here.
```suggestion
std::unique_ptr<LocalInteractionBase<traitsT>>
```

ToruNiina commented on test/core/test_directional_contact_interaction.cpp:
```diff
@@ -0,0 +1,230 @@
+              const auto rot_y = uni(mt) * pi;
+              const auto rot_z = uni(mt) * pi;
+
+              matrix33_type rotm_x(1.0,             0.0,              0.0,
```
const?

ToruNiina commented on test/core/test_directional_contact_interaction.cpp:
```diff
@@ -0,0 +1,230 @@
+              matrix33_type rotm_x(1.0,             0.0,              0.0,
+                                   0.0, std::cos(rot_x), -std::sin(rot_x),
+                                   0.0, std::sin(rot_x),  std::cos(rot_x));
+              matrix33_type rotm_y( std::cos(rot_y), 0.0,  std::sin(rot_y),
```
const?

ToruNiina commented on test/core/test_directional_contact_interaction.cpp:
```diff
@@ -0,0 +1,230 @@
+                                   0.0,              0.0, 1.0);
+
+              const matrix33_type rotm = rotm_x * rotm_y * rotm_z;
+              for(std::size_t idx=0; idx<sys.size(); ++idx)
```
const?

ToruNiina reviewed (changes requested):
I have some general comments.

1. It's too large (+1219!). At least, you could split this into the following 3 distinct PRs.
- update toml11
  - because this PR does not use the new feature in toml11, this is independent from this PR.
  - also, toml11 does not have any new release. pinning to the head is discouraged.
- add UniformPotential
- add DirectionalContactInteraction  

For ease of review, please, make PR as small as possible henceforth. Do NOT mix topics.

2. `const` by default. Non-const values easily cause bugs. Always use `const` if a value will not be modified. Leaving stuff non-const means that the value will be modified or moved out later. Always put `const` to your variable and remove if modification is required. 

3. I don't force you to use my indentation style, but at least you need to be consistent with yourself. The details like where to put a space, where to add a newline, etc. are important, actually. Keep your code clean, short and easy to read at a glance. Most of the semantic mistakes can be avoided by making code readable.

ToruNiina commented:
PS: Send a PR to Mjolnir-MD. Automatic code review tool is activated there.

ToruNiina commented on book/ja/refs/DirectionalContactInteraction.md:
```diff
@@ -0,0 +1,55 @@
+margin             = 0.5 # relative length to longest cutoff
+parameters         = [
+    # 各ポテンシャルそれぞれについて、要求されるパラメータを設定します。
+    {indices = [0, 1, 2, 3], angle1_params = { v0 = ... }, angle2_params = { v0 = ... }, contact_params = { v0 = ... }},
```
Do we really need `_params` suffix in `parameters` section? It is already clear that parameters are defined here, isn't it? I think `angle1 = {v0 = ...}, angle2 = {v0 = ...}, contact = {...}` is okay. Since this interaction takes 3 distinct potential parameters, lines would be long. There is no space to put redundant suffix.

ToruNiina commented on mjolnir/input/read_local_potential.hpp:
```diff
@@ -362,6 +385,48 @@ read_local_potentials(const toml::value& local)
+         typename angle2_potentialT, typename contact_potentialT>
+std::vector<std::tuple<std::array<std::size_t, N>, angle1_potentialT,
+                       angle2_potentialT, contact_potentialT>>
+read_directional_contact_potentials(const toml::value& local)
```
This function is too specific to call from other interactions. Is there a reason to split it from `read_directional_contact_interaction` stuff?

yutakasi634 replied:
The process corresponds to this function is implemented in read_directional_contact_iteraction more than once. Considering the amount of code for this process, first of all, I thought splitting code is better for readability.

ToruNiina replied:
Okay, but still, I feel this function is too specific. Isn't it good to move it to `read_local_interaction.hpp` because this function does not make sense without knowledge of the interaction.

ToruNiina commented on mjolnir/interaction/local/DirectionalContactInteraction.hpp:
```diff
@@ -0,0 +1,390 @@
+    return;
+  }
+
+  const real_type max_cutoff_length() const noexcept
```
meaningless `const`.
```suggestion
  real_type max_cutoff_length() const noexcept
```

ToruNiina commented on mjolnir/input/read_local_interaction.hpp:
```diff
@@ -122,6 +123,136 @@ read_contact_interaction(const std::string& kind, const toml::value& local)
+read_angle_in_directional_contact_interaction(
+    const std::string& kind, const toml::value& local, std::vector<std::string>&)
+{
+  return read_contact_in_directional_contact_interaction<traitsT, PotentialTs...>(kind, local);
```
You can put branching for contact potential in `read_contact_in_directional_contact_interaction` here, no?

ToruNiina reviewed (changes requested):
I made some other comments.

yutakasi634 committed:
refactor: fix intentation

yutakasi634 committed:
refactor: change the label of redundant include guard

yutakasi634 committed:
refactor: add lacking const to type declaration

yutakasi634 committed:
refactor: add referance to type declaration

yutakasi634 committed:
book: fix some typo

yutakasi634 committed:
refactor: spacing in the example input file

yutakasi634 committed:
refactor: unify coding style

yutakasi634 committed:
refactor: change argument of read_directional_contact_interaction

yutakasi634 committed:
refactor: fix current margin update timing

yutakasi634 committed:
refactor: change const to constexpr and rearrange indentation around this

yutakasi634 committed:
refactor: change initializing method of system position

yutakasi634 committed:
refactor: remove unnecessary typename keyword

yutakasi634 committed:
refactor: change initialization of array to aggregate initialization

yutakasi634 committed:
fix: change specification form of param for directional contact interaction

yutakasi634 committed:
fix: remove extra const keyword

yutakasi634 committed:
refactor: simplify recursion in read directional contact interaction

yutakasi634 commented:
If all the conversations resolved, I will make new PR to Mjolnir-MD and close this PR.

yutakasi634 committed:
test: adjust test to new parameter specifying style

yutakasi634 committed:
fix: enable to set the margin of directional contact interaction

ToruNiina commented on book/ja/refs/DirectionalContactInteraction.md:
```diff
@@ -0,0 +1,55 @@
+margin             = 0.5 # relative length to longest cutoff
+parameters         = [
+    # 各ポテンシャルそれぞれについて、要求されるパラメータを設定します。
+    {indices = [0, 1, 2, 3], ... }, angle1 = { v0 = ... }, angle2 = { v0 = ... }, contact = { v0 = ... }},
```
```suggestion
    {indices = [0, 1, 2, 3], angle1 = { v0 = ... }, angle2 = { v0 = ... }, contact = { v0 = ... }},
```

ToruNiina commented on mjolnir/input/read_local_interaction.hpp:
```diff
@@ -122,6 +124,108 @@ read_contact_interaction(const std::string& kind, const toml::value& local)
+    sizeof...(PotentialTs) < 2, std::unique_ptr<LocalInteractionBase<traitsT>>
+    >::type
+read_directional_contact_interaction(
+    const std::string& kind, const toml::value& local, std::vector<std::string>&& angle_potential_keys = {"angle2", "angle1"})
```
I think it is good to pass the last argument explicitly in the line 537. This argument is important, so I don't think it is a good way to make it "optional".

ToruNiina commented on test/core/test_directional_contact_interaction.cpp:
```diff
@@ -0,0 +1,230 @@
+                                           0.0, std::sin(rot_x),  std::cos(rot_x));
+                const matrix33_type rotm_y(std::cos(rot_y), 0.0,  std::sin(rot_y),
+                                           0.0,             1.0,              0.0,
+                                           -std::sin(rot_y), 0.0,  std::cos(rot_y));
```
```suggestion
                                          -std::sin(rot_y), 0.0,  std::cos(rot_y));
```

ToruNiina commented on mjolnir/input/read_local_potential.hpp:
```diff
@@ -362,6 +362,49 @@ read_local_potentials(const toml::value& local)
+    MJOLNIR_LOG_INFO("as", N, "-body interaction");
+
+    using indices_type = std::array<std::size_t, N>;
+    using indices_potentials_tuple_type =
```
```suggestion
    using indices_potentials_tuple_type = std::tuple<
```

yutakasi634 committed:
book: fix directional contact interactin paramter description

yutakasi634 committed:
refactor: move specific read func to read interaction file

yutakasi634 committed:
refactor: remove optional arg and pass explicitly

remove optional arg of read_direcitonal_contact_interaction func and pass this arg explicitly in read local interaction.

yutakasi634 committed:
refactor: rearrange indentation

yutakasi634 commented:
Can I close and remake PR for Mjolnir-MD?

ToruNiina reviewed (approved):
LGTM.
